controllers/class.tapestry-controller.php

```diff
@@ -0,0 +1,337 @@
+        $node = $tapestryNodeController->save($node);
+
+        if (is_wp_error($node)) {
+            return $node;
+        }
+
+        $tapestry = get_post_meta($this->postId, 'tapestry', true);
+
+        if (!isset($tapestry->nodes)) {
+            $tapestry->nodes = [];
+        }
+
+        array_push($tapestry->nodes, $node->id);
+
+        if (empty($tapestry->rootId)) {
+            $tapestry->rootId = $tapestry->nodes[0];
+        }
+
+        update_post_meta($this->postId, 'tapestry', $tapestry);
+
+        return $node;
+    }
+
+    /**
+     * Add a new link
+     * 
+     * @param  Object   $link   Tapestry link
+     * 
+     * @return  Object  $link   Tapestry link
+     */
+    public function addLink($link)
+    {
+        $tapestryLinkController = new TapestryLinkController($this->postId);
+        $link = $tapestryLinkController->save($link);
+
+        return $link;
+    }
+
+    /**
+     * Add a new group
+     * 
+     * @param   Object  $group   Tapestry group
+     * 
+     * @return  Object  $group   Tapestry group
+     */
+    public function addGroup($group)
+    {
+        $tapestryGroupController = new TapestryGroupController($this->postId);
+        $group = $tapestryGroupController->save($group);
+
+        if (is_wp_error($group)) {
+            return $group;
+        }
+
+        $tapestry = get_post_meta($this->postId, 'tapestry', true);
+
+        if (!isset($tapestry->groups)) {
+            $tapestry->groups = [];
+        }
+
+        array_push($tapestry->groups, $group->id);
+
+        update_post_meta($this->postId, 'tapestry', $tapestry);
+
+        return $group;
+    }
+
+    /**
+     * Get the node controller with associated node meta ID
+     * 
+     * @param   Number  $nodeMetaId node meta ID
+     *
+     * @return  Object  $node       node controller
+     */
+    public function getNode($nodeMetaId)
+    {
+        return new TapestryNodeController($this->postId, $nodeMetaId);
+    }
+
+    /**
+     * Get the group controller
+     * 
+     * @return  Object  $group  group controller
+     */
+    public function getGroups()
+    {
+        return new TapestryGroupController($this->postId);
+    }
+
+    /**
+     * Get the link controller
+     * 
+     * @return  Object  $link   link controller
+     */
+    public function getLinks()
+    {
+        return new TapestryLinkController($this->postId);
+    }
+
+    /**
+     * Get the setting controller
+     * 
+     * @return  Object  $setting   setting controller
+     */
+    public function getSettings()
+    {
+        return new TapestrySettingController($this->postId);
+    }
+
+    private function _addTapestry($tapestry)
+    {
+        $this->postId = TapestryHelpers::updatePost($tapestry, 'tapestry');
+
+        $tapestry->links = [];
+        $tapestry->nodes = [];
+        $tapestry->groups = [];
+
+        update_post_meta($this->postId, 'tapestry', $tapestry);
```

Comment:
I recommend creating a private method for gathering all the info (from the class private properties) needed to go into this `$tapestry`.